### ISSUE TITLE
Some questions about source code.

### ISSUE THREAD

**Issue by shinima** [redux-saga/redux-saga#1399]
Question 1:  I found that `mainTask.isMainRunning` is set to `false` [here](https://github.com/redux-saga/redux-saga/blob/d98344e4fd6a7be561bd26f58d5c831e3b14c1b8/packages/core/src/internal/proc.js#L314) and [here](https://github.com/redux-saga/redux-saga/blob/d98344e4fd6a7be561bd26f58d5c831e3b14c1b8/packages/core/src/internal/proc.js#L321), but I cannot find where `mainTask.isMainRunning` is used. I am wandering whether it is a mistake.

Question 2: [Here](https://github.com/redux-saga/redux-saga/blob/d98344e4fd6a7be561bd26f58d5c831e3b14c1b8/packages/core/src/internal/proc.js#L301) that a `CHANNEL_END` will make the iterator jump to finally block without cancelling its attached forks. I know its behavior, but I cannot understand its intention. Can you explain why? An example will be grateful.

**Comment by Andarist** [comment]
1. probably some leftover, would u like to send a PR removing those assignments? OTOH there is a chance that we should set `isRunning` to false there, but it would require more thorough investigation

2. truth to be told - im not sure about this, `CHANNEL_END` is a somewhat hacky solution for SSR - I guess the intention is that every started task should be completed until it hits `take` effect which ends this particular task with `CHANNEL_END` but not necessarily its forks - they should end in the same way (by getting `take` effect)

**Comment by shinima** [comment]
1. I have made a PR 😄 
2. Let's just leave it alone.

And... Question 3:

[Here](https://github.com/redux-saga/redux-saga/blob/7b415a2408f9fb9002abae1112dce3b6bb5f6ed6/packages/core/src/internal/proc.js#L774-L778) we are using properties of iterator to track the parent-task status, but actually [_the parent task is the  aggregation of the main tasks + all its forked tasks_](https://github.com/redux-saga/redux-saga/blob/master/packages/core/src/internal/proc.js#L62). 
So, in current implementation, there are confusing situations if we run the following code:

```JavaScript
function* genFn() {
  yield fork(backgroundTask)
  // mainTask.isRunning is set to false here
  // But since the backgroundTask is still running, so the parent-task is still running
  // We are using iterator._isRunning to track parent-task so iterator._isRunning is true.
  // It's confusing that:
  //    mainTask.isRunning === false && iterator._isRunnig === true
}
```

I think its better to use `task._isRunning / task._isCancelled ...` to track the status of parent-task. I will make another PR for this.

Point out the mistakes if I am wrong. And if it is necessary, I apologize for my poor English.

**Comment by Andarist** [comment]
I agree this is somewhat confusing, I'm not sure though if we should track those flags on `task`. Need to investigate this to evaluate out options and relations between those things, thanks for the PRs! I'm quite busy at the moment though, so I won't have time to tackle this now, but I will for sure get back to it whenever I have some spare time.

**Comment by restrry** [comment]
@shinima there is an example with similar setup https://codesandbox.io/s/7z8pl8jl11
@Andarist  correct me if I'm wrong
if `backgroundTask` is detached process  then `genFn` will have `isRunning === false` state 
if `backgroundTask` is non-detached then  `genFn` tracks `backgroundTask` and waits for it to finish.

**Comment by shinima** [comment]
@restrry Thanks for answering my questions.  
What I mean is that there are confusing situations when `mainTask.isRunning === false && iterator._isRunnig === true` holds. (`mainTask` is an implementation detail in redux-saga.)
In my opinion, it is better to refactor these codes and use `task._isRunning / task._isCancelled ...` to track the status. 
However, it is a refinement with low-priority, and I found that my proposal(#1404) is not a good one. I will investigate better options.

**Comment by Andarist** [comment]
I think the problem is that those are in fact 2 different states - when the whole task is "active"/running and when `mainTask` is finished. Iterator got exhausted but there still some child tasks running.

Both flags have different meaning and both are used for other things, I agree though that naming them both similarly is unfortunate and confusing, we could rename one of them to make it more obvious what's going on.

**Comment by shinima** [comment]
@Andarist Yes, I spent several hours figuring out this confusing names.
Maybe using `iterator._isTaskRunning / iterator._isTaskCancelled / iterator._taskResult ...` is a better option. This makes the meaning of these properties explicitly. How do you think?
I am glad to make a PR.
